src/common/framework/tree.ts

```diff
@@ -1,82 +1,306 @@
-      partial += parts[i] + '/';
-      yield partial;
+export interface FilterResultTreeLeaf {
+  readonly query: TestQuerySingleCase;
+  readonly run: RunFn;
+}
+
+export type FilterResultTreeNode = FilterResultSubtree | FilterResultTreeLeaf;
+
+export class FilterResultTree {
+  readonly root: FilterResultSubtree;
+
+  constructor(root: FilterResultSubtree) {
+    this.root = root;
+  }
+
+  iterateCollapsedQueries(): IterableIterator<TestQuery> {
+    return FilterResultTree.iterateSubtreeCollapsedQueries(this.root);
+  }
+
+  iterateLeaves(): IterableIterator<FilterResultTreeLeaf> {
+    return FilterResultTree.iterateSubtreeLeaves(this.root);
+  }
+
+  toString(): string {
+    return FilterResultTree.subtreeToString('(root)', this.root, '');
+  }
+
+  static *iterateSubtreeCollapsedQueries(
+    subtree: FilterResultSubtree
+  ): IterableIterator<TestQuery> {
+    for (const [, child] of subtree.children) {
+      if ('children' in child && !child.collapsible) {
+        yield* FilterResultTree.iterateSubtreeCollapsedQueries(child);
+      } else {
+        yield child.query;
+      }
     }
-    // Path ends in '/' (so is a README).
-    if (parts[parts.length - 1] === '') {
-      return;
+  }
+
+  static *iterateSubtreeLeaves(
+    subtree: FilterResultSubtree
+  ): IterableIterator<FilterResultTreeLeaf> {
+    for (const [, child] of subtree.children) {
+      if ('children' in child) {
+        yield* FilterResultTree.iterateSubtreeLeaves(child);
+      } else {
+        yield child;
+      }
     }
   }
-  yield path + terminator;
-}
 
-export function treeFromFilterResults(
-  log: Logger,
-  listing: IterableIterator<TestFilterResult>
-): FilterResultTreeNode {
-  function getOrInsert(n: FilterResultTreeNode, k: string): FilterResultTreeNode {
-    const children = n.children!;
-    if (children.has(k)) {
-      return children.get(k)!;
+  static subtreeToString(name: string, tree: FilterResultTreeNode, indent: string): string {
+    const collapsible = 'run' in tree ? '>' : tree.collapsible ? '+' : '-';
+    let s =
+      indent +
+      `${collapsible} ${JSON.stringify(name)} => ` +
+      `${tree.query}        ${JSON.stringify(tree.query)}`;
+    if ('children' in tree) {
+      if (tree.description !== undefined) {
+        s += indent + `\n    | ${JSON.stringify(tree.description)}`;
+      }
+
+      for (const [name, child] of tree.children) {
+        s += '\n' + FilterResultTree.subtreeToString(name, child, indent + '  ');
+      }
     }
-    const v = { children: new Map() };
-    children.set(k, v);
-    return v;
+    return s;
   }
+}
 
-  const tree = { children: new Map() };
-  for (const f of listing) {
-    const files = getOrInsert(tree, f.id.suite + ':');
-    if (f.id.path === '') {
-      // This is a suite README.
-      files.description = f.spec.description.trim();
+// TODO: Consider having subqueriesToExpand actually impact the depth-order of params in the tree.
+export async function loadTreeForQuery(
```

Comment:
Overall I worry that this fully expands the tree in memory: every test case of every test has a node, which could get big, and also each leaf node contains the full file / test path. I don't think this can be addressed now, but it'd be nice to remove some of the duplication or load cases lazily at some point because otherwise the runners might start taking a long time to start.

Most of the refactors related to queries and other wouldn't be affected so I think it is fine to do it later down the line.

Reply:
Added TODO